Submenu blog post (#6143)
adobe/react-spectrum: packages/dev/docs/pages/blog/creating-a-pointer-friendly-submenu-experience.mdx

<SubmenuSafeArea
  role="img"
  aria-label="Diagram showing safe area triangle drawn between pointer and submenu" />

## Arctangent

We can measure the angle at which the pointer moved by using the 2-argument arctangent, or [atan2](https://en.wikipedia.org/wiki/Atan2). If we provide our pointer’s delta x and delta y values as arguments, we’ll get the angle (in radians) from the previous pointer position to the current pointer position.

<SubmenuAtan2
  role="img"
  aria-label="Diagram showing atan2 value as angle between the X axis and some line on a cartesian plane" />

Now we can use the atan2 function to measure the angles formed by three separate lines:

* **Θ<sub>top</sub>** : Angle formed by the line from the previous pointer position to the **top inside corner** of the submenu
* **Θ<sub>bottom</sub>** : Angle formed by the line from the previous pointer position to the **bottom inside corner** of the submenu
* **Θ<sub>pointer</sub>** : Angle formed by the line from the previous pointer position to the **current pointer** position (delta)

> <SubmenuAtan2FromPointer

Reviewer: <img width="151" alt="Screenshot 2024-04-05 at 1 21 16 PM" src="https://github.com/adobe/react-spectrum/assets/44729383/fd5e7ab7-52dd-469a-b654-aa1609efc690">

would it be possible to add a border or something to the previous pointer? it's kinda hard to see in dark mode

  role="img"
  aria-label="Diagram showing triangles formed between pointer and various points on submenu" />

If the pointer’s delta angle is **between** the top and bottom angles, we know the user is moving their pointer in the direction of the submenu.

<p style={{textAlign: 'center', fontSize: 'xx-large'}}>Θ<sub>top</sub> > Θ<sub>pointer</sub> > Θ<sub>bottom</sub></p>

> ## Time

Reviewer: I'm not sure if "Time" is the best title for this section. Maybe something like "Pointer Speed" since you mention speed being one of the ways of predicting user intent tho I'm not sold on that title either lol. Will ponder...

Author: I like "Pointer Speed" better. In fact, I think I'll rename "Arctangent" to "Pointer Direction" so they're both inline with the earlier section.

> We also want to consider the pointer’s movement over time when predicting the user’s intent. A user typically increases the **speed** of their pointer when moving towards the submenu, then decreases it as they reach their target. The user might change their mind on the way, however, and slow down to a stop or keep browsing other options in the parent menu. If their pointer speed dips below a certain threshold before arriving at the submenu, we could assume they're no longer intending to go to the submenu.

Reviewer 2: ```suggestion
In order to predict the user’s intent, we also need to know the pointer's velocity. A user typically accelerates their pointer when moving towards the submenu, then decelerates as they reach their target.

The user can slow down to a stop or continue browsing other options in the parent menu. If their pointer speed dips below a certain threshold before arriving at the submenu, we could assume they're no longer intending to go there or the submenu was opened accidentally.
```

Reviewer 2: I'm still not quite happy with my suggestion, it's hard to tell how this relates to speed as opposed to angle: "continue browsing other options in the parent menu"

Author: I was more talking about stopping your pointer movement to read the options in the submenu (after you've already left the trigger item). I can definitely reword that.

> We could check the pointer’s speed throughout its journey with our knowledge of what speeds to expect at the different stages along the way. We can even take it a step further by using our knowledge of the [Accot–Zhai steering law](https://en.wikipedia.org/wiki/Steering_law) to predict the time it may take the user to reach the target based on the width of the “tunnel” to the submenu.

Reviewer: Could you add an additional space between each paragraph. I assume this is what was intended based on the format in this file but in the docs it's just rendering as one giant paragraph and I think it would be nice if it was broken up a little bit.

> Alternatively, we can go with a simpler solution and implement a **timeout**. If the pointer hasn’t moved after a certain amount of time, we can assume they are no longer intending to go to the submenu. Since users with motor impairments may take more time to move their pointer, we should lean towards using a larger timeout value. Although our timeout solution is more simple than incorporating the pointer’s speed, we found that it still results in a good user experience.

Reviewer 2: Which one did we use? the tunnel or the simple solution?

I think we're using the simple one? Only talk about that one, and you can use an aside or footnote or something less catching for the tunnel method "additional reading, here is link"

Right now, they're competing for my attention.

## Fault Tolerance

> Since our users aren’t perfect, we want to build in some fault tolerance in case their actions are outside the scope we have defined so far, but not by enough to invalidate their intent. Here are some ways we can do that:

Reviewer 2: we **can** do that? or we **did**?

Reviewer 2: ```suggestion
Since our users are human, we want to build in some fault tolerance, but not so much that we invalidate their intent. Here are some ways we can do that:
```

* Widen our range of allowed angles: Add radians to our top and bottom angles.

> * Allow N invalid movements: Require that the user make two (or some larger constant) consecutive invalid pointer movements before closing the submenu. Users who experience [tremors](https://en.wikipedia.org/wiki/Tremor) may involuntarily move their pointer in other directions, so it is important to include this feature.

Reviewer 2: ```suggestion
* Allow N invalid movements: Require that the user make `N` consecutive invalid pointer movements before closing the submenu. Users who experience [tremors](https://en.wikipedia.org/wiki/Tremor) may involuntarily move their pointer in other directions, so it is important to include this feature.
```

## Optimizations

We can introduce a few performance optimizations:

> * Throttle: Most devices refresh their screens 60 times per second. This means that we don't need to do these measurements more frequently than every 16 ms (1 second / 60 = 16.66 ms). We can keep timestamps and skip checks if enough time hasn't passed. We can also experiment with doing checks even less frequently while still maintaining a good user experience. Lowering the sample rate would also provide more accurate predictions for users who experience tremors.

Reviewer: I think you could make "throttle" and "avoid unnecessary work" bold

and same goes for the "Alternatives" section

> * Throttle: Most devices refresh their screens 60 times per second. This means that we don't need to do these measurements more frequently than every 16 ms (1 second / 60 = 16.66 ms). We can keep timestamps and skip checks if enough time hasn't passed. We can also experiment with doing checks even less frequently while still maintaining a good user experience. Lowering the sample rate would also provide more accurate predictions for users who experience tremors.

Reviewer 2: ```suggestion
* Throttle: Most devices refresh their screens 60 times per second. This means that we don't need to do these measurements more frequently than every 16 ms (1 second / 60 = 16.66 ms). We can also experiment with doing checks even less frequently while still maintaining a good user experience. For instance, lowering the sample rate may provide more accurate predictions for users who experience tremors.
```

## Pointer Speed

In order to predict the user’s intent, we also need to know the pointer's speed. Here are some things we know about how users move their pointer:

* Users typically accelerate their pointer when moving towards the submenu, then decelerate as they reach their target. 
* Users sometimes stop or slow down their pointer to browse options in the submenu.
* Users tend to move their pointer more quickly if they have a larger distance to cover (see [Fitts's Law](https://en.wikipedia.org/wiki/Fitts%27s_law)).
* Users tend to move their pointer more quickly if they have a larger "tunnel" to navigate through (see [Steering Law](https://en.wikipedia.org/wiki/Steering_law)).

We could check the pointer’s speed continuously and use the above knowledge to predict the user’s intent.

Alternatively, we could simply use a **timeout**, where if the pointer hasn’t moved after a certain amount of time and is no longer over the submenu's parent menu item, we assume they are no longer intending to go to the submenu.
Since users with motor impairments may take more time to move their pointer, we should lean towards using a larger timeout value.

> Although the timeout solution is simpler than tracking the pointer’s speed, we found that it still resulted in a good user experience, so we proceeded with this approach.

Author: I think we should keep the parts about tracking speed to show our thought process in building this, but maybe I can be more clear about why the timeout can be a good alternative.

> Although the timeout solution is simpler than tracking the pointer’s speed, we found that it still resulted in a good user experience, so we proceeded with this approach.

Author: I think 'speed' still works since it means distance covered over time, and the timeout represents covering no distance over a specific time.

## Fault Tolerance

Since our users are human, we want to build in some fault tolerance, but not so much that we invalidate their intent. Here are some ways we did that:
* **Widen our range of allowed angles**: We added 15 degrees of tolerance to the top and bottom angles. After testing different values, we found that this created the best experience.
* **Allow invalid movements**: We require that two consecutive invalid pointer movements be made before closing the submenu. Users who experience [tremors](https://en.wikipedia.org/wiki/Tremor) may involuntarily move their pointer in other directions, so it is important to include this feature.  

## Optimizations

We can introduce a few performance optimizations:

* **Throttle**: Most devices refresh their screens 60 times per second. This means that we don't need to do these measurements more frequently than every 16 ms (1 second / 60 = 16.66 ms). We can also experiment with doing checks even less frequently while still maintaining a good user experience. For instance, lowering the sample rate may provide more accurate predictions for users who experience tremors.

> * **Avoid unnecessary work**: We don’t need to track pointer movement if a submenu isn’t open, and we can stop tracking pointer movement once the pointer reaches the submenu. We can also check the pointer’s delta x before calculating any angle values. If the pointer is moving in the opposite direction of the submenu, there's no need to calculate and compare angles. We can also check our pointer event's [pointerType](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerType) property and ignore the event if it is type 'pen' or 'touch'.

Author: I agree, I think this could be broken up.

* **Check if the point is within a triangle**: We could define a triangle from the previous pointer position to the top inside corner of the submenu and the bottom inside of the submenu. This method is outlined by Ben Kamens in [Breaking down Amazon’s mega dropdown](https://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown).
* **Render an invisible triangle over the parent menu**: We could use an absolute-positioned HTML element and use [clip-path](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path) to give it a triangle shape. There is an excellent post by Andreas Eldh called [Invisible Details](https://medium.com/linear-app/invisible-details-2ca718b41a44) that walks through how to implement this. Similarly, we could draw an [SVG](https://developer.mozilla.org/en-US/docs/Web/SVG) instead, as outlined by Costa Alexoglou in [Better Context Menus With Safe Triangles](https://www.smashingmagazine.com/2023/08/better-context-menus-safe-triangles/). This method is also mentioned in [Building like it's 1984: A comprehensive guide to creating intuitive context menus](https://height.app/blog/guide-to-build-context-menus) by Michael Villar.

## Conclusion

> We hope this post has been helpful in understanding how we approached the problem of predicting user intent when using submenus. We are excited to see how you use submenus in your own projects. You can see this feature in action in the [React Spectrum Menu](https://react-spectrum.adobe.com/react-spectrum/Menu.html#submenus) and [React Aria Menu](https://react-spectrum.adobe.com/react-aria/Menu.html#submenus).

Reviewer 3: `problem of predicting user intent when using submenus` Should this be more explicate about user intent when using a pointer to open a submenu and interact with it or let it close?

Author: Definitely agree, I'll reword this.

OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import SubmenuSafeArea from '../assets/submenu-safe-area.svg';
import SubmenuAtan2 from '../assets/submenu-atan2.svg';
import SubmenuAtan2FromPointer from '../assets/submenu-atan2-from-pointer.svg';

import {BlogPostLayout} from '@react-spectrum/docs';
export default BlogPostLayout;

---
keywords: [react aria, react spectrum, react, spectrum, interactions, submenu, pointer]
description: We are excited to announce support for submenus in the latest release of React Spectrum, React Aria, and React Aria Components! In the process of adding this feature, we found ourselves solving some unique challenges while working to make submenus user-friendly and accessible across an array of devices and input types. In doing so, we wanted to share our thought process in solving one of the challenges we faced along the way.
date: 2024-04-03
author: '[Reid Barber](https://github.com/reidbarber)'
image: ../assets/submenu-safe-area.png
---

# Creating a pointer-friendly submenu experience

> We are excited to announce a beta release of submenus in the latest release of React Spectrum, React Aria, and React Aria Components! In the process of adding this feature, we found ourselves solving some unique challenges while working to make submenus user-friendly and accessible across an array of devices and input types. In doing so, we wanted to share our thought process in solving one of the challenges we faced along the way.

Reviewer 4: ```suggestion
We are excited to announce support for submenus in the latest release of React Spectrum, React Aria, and React Aria Components! In the process of adding this feature, we found ourselves solving some unique challenges while working to make submenus user-friendly and accessible across an array of devices and input types. In doing so, we wanted to share our thought process in solving one of the challenges we faced along the way.
```

Reviewer 4: Can we also link to the docs here (e.g. on React Spectrum and React Aria Components)? Not sure about React Aria since we have no docs... I know we link at the end but might be nice to do it at the top too so people can try it right away.

Author: Maybe I'll update this to React Spectrum and React Aria.

Author: We'll also have a demo on the React Aria homepage now, so I think I'll also link to that at the end.

> ## The Shortcut Problem

Reviewer 4: Why is it called the "shortcut" problem? You didn't mention shortcut again so I wasn't sure.

Author: I made that up. Wasn't sure of a better section heading. I'll update that section to reference it.

Author: Renamed this to `The Shortest Path` to make it more clear.

* **Track movements only when necessary**: We can start tracking pointer movements when the submenu opens and stop tracking when the submenu closes, or the pointer reaches the submenu. 
* **Calculate angles only when necessary**: If the pointer is moving in the opposite direction of the submenu, there's no need to calculate and compare angles. We can check this by comparing the pointer’s delta x to the submenu’s closest edge's x.
* **Check the pointer event type**: We can check the pointer event's [pointerType](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerType) property and ignore the event if it is type 'pen' or 'touch'.

## Alternatives

Let's compare our approach to a few other methods:

* **Timeout only**: We could use just a timeout instead of incorporating the direction the pointer moves. The downside of this is that moving the pointer vertically between parent menu items would cause delayed interactions that could be unpleasant or unexpected for the user.
* **Delay before opening submenus**: We could introduce a delay before opening each submenu, but that would introduce a similar negative user experience as the timeout-only method described above.

> * **Check if the point is within a triangle**: We could define a triangle from the previous pointer position to the top inside corner of the submenu and the bottom inside of the submenu. This method is outlined by Ben Kamens in [Breaking down Amazon’s mega dropdown](https://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown).

Reviewer 4: Can we add something here about how this differs from our approach and why we didn't go with it? It seem very similar.

Author: Yeah, based on their wording I assumed they were doing something like a [point-in-polygon](https://en.wikipedia.org/wiki/Point_in_polygon) method, but looks like they are actually using the slope formula. This is similar to our solution, but ours ends up comparing values in radians via [Math.atan2](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/atan2).

> * **Render an invisible triangle over the parent menu**: We could use an absolute-positioned HTML element and use [clip-path](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path) to give it a triangle shape. There is an excellent post by Andreas Eldh called [Invisible Details](https://medium.com/linear-app/invisible-details-2ca718b41a44) that walks through how to implement this. Similarly, we could draw an [SVG](https://developer.mozilla.org/en-US/docs/Web/SVG) instead, as outlined by Costa Alexoglou in [Better Context Menus With Safe Triangles](https://www.smashingmagazine.com/2023/08/better-context-menus-safe-triangles/). This method is also mentioned in [Building like it's 1984: A comprehensive guide to creating intuitive context menus](https://height.app/blog/guide-to-build-context-menus) by Michael Villar.

Reviewer 4: Might be worth adding something about why we switched away from this approach here too.
